src/c2ffi/c2ffi.lisp

```diff
@@ -0,0 +1,163 @@
+      ;; Invoke c2ffi to emit macros into TMP-MACRO-FILE
+      (when (run-program* *c2ffi-executable* (list* (namestring input-header-file)
+                                                    "-D" "null"
+                                                    "-M" (namestring tmp-macro-file)
+                                                    (append arch sys-include-paths))
+                          :output *standard-output*
+                          :ignore-error-status ignore-error-status)
+        ;; Write a tmp header file that #include's the input file and the macros file.
+        (uiop:with-temporary-file (:stream tmp-include-file-stream
+                                   :pathname tmp-include-file
+                                   :keep *trace-c2ffi*)
+          (format tmp-include-file-stream "#include \"~A\"~%" input-header-file)
+          (format tmp-include-file-stream "#include \"~A\"~%" tmp-macro-file)
+          :close-stream
+          ;; Invoke c2ffi again to generate the final output.
+          (run-program* *c2ffi-executable* (list* (namestring tmp-include-file) "-o" (namestring output-spec-path)
+                                                  (append arch sys-include-paths))
+                        :output *standard-output*
+                        :ignore-error-status ignore-error-status))))))
+
+(defun spec-path (base-name &key version (arch (local-arch)))
+  (check-type base-name pathname)
+  (make-pathname :defaults base-name
+                 :name (strcat (pathname-name base-name)
+                               (if version
+                                   (strcat "-" version)
+                                   "")
+                               "."
+                               arch)
+                 :type "spec"))
+
+(defun find-local-spec (base-name &key (otherwise nil otherwise?))
+  (let* ((spec-path (spec-path base-name))
+         (probed (probe-file spec-path)))
+    (if probed
+        spec-path
+        (progn
+          (unless otherwise?
+            (error "c2ffi spec file not found for base name ~S" base-name))
+          otherwise))))
+
+(defun ensure-spec-file-exists (header-file-path &key exclude-archs sys-include-paths version)
+  (multiple-value-bind
+        (h-name m-name)
+      (find-local-spec header-file-path :otherwise nil)
+    (if h-name
+        (values h-name m-name)
+        (let ((local-arch (local-arch)))
+          (unless (c2ffi-executable-available?)
+            (error "No spec found for ~S on arch '~A' and c2ffi not found"
+                   header-file-path local-arch))
+          (generate-spec-with-c2ffi header-file-path
+                                    (spec-path header-file-path :arch local-arch :version version)
+                                    :arch local-arch
+                                    :sys-include-paths sys-include-paths)
+          ;; also run c2ffi for other architectures, but tolerate failure
```

Comment:
What's the goal of running c2ffi for multiple architectures? Is it help maintainers populate their library of pre-built specs?